3-enrich/stream-enrich/core/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/utils.scala

```diff
@@ -49,4 +70,98 @@ object utils {
   implicit val uuidProvider: UUIDProvider[Id] = new UUIDProvider[Id] {
     override def generateUUID: Id[UUID] = UUID.randomUUID()
   }
+
+  def getAWSCredentialsProvider(creds: Credentials): Either[String, AWSCredentialsProvider] = {
+    def isDefault(key: String): Boolean = key == "default"
+    def isIam(key: String): Boolean = key == "iam"
+    def isEnv(key: String): Boolean = key == "env"
+
+    for {
+      awsCreds <- creds match {
+        case NoCredentials => "No AWS credentials provided".asLeft
+        case _: GCPCredentials => "GCP credentials provided".asLeft
+        case c: AWSCredentials => c.asRight
+      }
+      provider <- awsCreds match {
+        case AWSCredentials(a, s) if isDefault(a) && isDefault(s) =>
+          new DefaultAWSCredentialsProviderChain().asRight
+        case AWSCredentials(a, s) if isDefault(a) || isDefault(s) =>
+          "accessKey and secretKey must both be set to 'default' or neither".asLeft
+        case AWSCredentials(a, s) if isIam(a) && isIam(s) =>
+          InstanceProfileCredentialsProvider.getInstance().asRight
+        case AWSCredentials(a, s) if isIam(a) && isIam(s) =>
+          "accessKey and secretKey must both be set to 'iam' or neither".asLeft
+        case AWSCredentials(a, s) if isEnv(a) && isEnv(s) =>
+          new EnvironmentVariableCredentialsProvider().asRight
+        case AWSCredentials(a, s) if isEnv(a) || isEnv(s) =>
+          "accessKey and secretKey must both be set to 'env' or neither".asLeft
+        case AWSCredentials(a, s) =>
+          new AWSStaticCredentialsProvider(new BasicAWSCredentials(a, s)).asRight
+      }
+    } yield provider
+  }
+
+  /**
+   * Create GoogleCredentials based on provided service account credentials file
+   * @param creds path to service account file
+   * @return
+   */
+  def getGoogleCredentials(creds: Credentials): Either[String, GoogleCredentials] =
+    for {
+      gcpCreds <- creds match {
+        case NoCredentials => "No GCP Credentials provided".asLeft
+        case _: AWSCredentials => "AWS credentials provided".asLeft
+        case c: GCPCredentials => c.asRight
+      }
+      credsPath = gcpCreds.creds
+      googleCreds <- {
+        if (Files.isRegularFile(Paths.get(credsPath)))
+          GoogleCredentials
+            .fromStream(new FileInputStream(credsPath))
+            .createScoped("https://www.googleapis.com/auth/cloud-platform")
+            .asRight
+        else "Provided Google Credentials Path isn't valid".asLeft
+      }
+    } yield googleCreds
+
+  /**
+   * Downloads an object from S3 and returns whether or not it was successful.
+   * @param uri The URI to reconstruct into a signed S3 URL
+   * @param targetFile The file object to write to
+   * @param provider necessary credentials to download from S3
+   * @return the download result
+   */
+  def downloadFromS3(
+    provider: AWSCredentialsProvider,
+    uri: URI,
+    targetFile: File
+  ): Either[Throwable, Int] = {
```

Comment:
Could be `Either[Throwable, Int]`, we don't need to return `Right(0)`, if it's a `Right` we know that it was successful

Reply:
The reason we return successful exit code after a potentially exception-throwing line actually comes from the first branch of `download()`. Please check the method and you'll see that in case uri scheme is `http(s)`, we use `!`:

```
def ! : Int // returns exit code
```

Both AWS and GCP SDKs' relevant methods return different types and both aren't `Int`.

Upstream in the codebase, `cacheFiles` method [uses](https://github.com/snowplow/snowplow/blob/master/3-enrich/stream-enrich/core/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/Enrich.scala#L285-L288) exit code semantic to interpret download failures.

I want to double check if you still think we should get rid of exit code usage altogether, which also means to update how `cacheFiles` interpret failure. Currently it doesn't make any use of failures (`Left` values) received from credential or download methods. @benjben

Reply:
Thanks for the details @oguzhanunlu.

> I say we get rid of exit code semantics and make use of thrown exceptions to interpret download failures, wdyt?

Totally with you!